src/main/java/io/cdap/plugin/gcp/dataplex/sink/DataplexBatchSink.java

```diff
@@ -0,0 +1,582 @@
+    try (DataplexServiceClient dataplexServiceClient = DataplexUtil.getDataplexServiceClient(config.getCredentials())) {
+      config.validateAssetConfiguration(collector, dataplexServiceClient);
+      config.validateServiceAccount(collector);
+      asset =
+        dataplexServiceClient.getAsset(AssetName.newBuilder().setProject(config.tryGetProject())
+          .setLocation(config.getLocation())
+          .setLake(config.getLake()).setZone(config.getZone()).setAsset(config.getAsset()).build());
+
+      if (config.getAssetType().equals(DataplexConstants.BIGQUERY_DATASET_ASSET_TYPE)) {
+        config.validateBigQueryDataset(context.getInputSchema(), context.getOutputSchema(), collector,
+          dataplexServiceClient);
+        prepareRunBigQueryDataset(context);
+      } else if (config.getAssetType().equals(DataplexConstants.STORAGE_BUCKET_ASSET_TYPE)) {
+        config.validateStorageBucket(collector);
+        prepareRunStorageBucket(context);
+      }
+    }
+  }
+
+  /**
+   * O/P Template parameters will get changed based on asset type. E.g. <StructuredRecord, NullWritable> for BQ Dataset
+   *
+   * @param input
+   * @param emitter
+   */
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<KeyValue<Object, Object>> emitter) {
+    if (this.config.getAssetType().equalsIgnoreCase(DataplexConstants.BIGQUERY_DATASET_ASSET_TYPE)) {
+      emitter.emit(new KeyValue<>(input, NullWritable.get()));
+    } else {
+      emitter.emit(new KeyValue<>(NullWritable.get(), input));
+    }
+  }
+
+  @Override
+  public void onRunFinish(boolean succeeded, BatchSinkContext context) {
+    if (this.config.getAssetType().equalsIgnoreCase(DataplexConstants.STORAGE_BUCKET_ASSET_TYPE)) {
+      emitMetricsForStorageBucket(succeeded, context);
+      return;
+    }
+
+    Path gcsPath = new Path(String.format("gs://%s", runUUID));
+    try {
+      FileSystem fs = gcsPath.getFileSystem(baseConfiguration);
+      if (fs.exists(gcsPath)) {
+        fs.delete(gcsPath, true);
+        LOG.debug("Deleted temporary directory '{}'", gcsPath);
+      }
+      emitMetricsForBigQueryDataset(succeeded, context);
+    } catch (IOException e) {
+      LOG.warn("Failed to delete temporary directory '{}': {}", gcsPath, e.getMessage());
+    } catch (Exception exception) {
+      LOG.warn("Exception while trying to emit metric. No metric will be emitted for the number of affected rows.",
+        exception);
+    }
+  }
+
+  /**
+   * prepare Run for BigQuery Dataset Asset. It will create necessary resources, and configuration methods
+   *
+   * @param context
+   * @throws Exception
+   */
+  private void prepareRunBigQueryDataset(BatchSinkContext context) throws Exception {
+    FailureCollector collector = context.getFailureCollector();
+    Credentials credentials = config.getCredentials();
+    String project = config.getProject();
+    String cmekKey = context.getArguments().get(GCPUtils.CMEK_KEY);
```

Comment:
Does it make sense to also add CMEK as a config `Encryption Key Name`  like we have in other GCP plugins?

Reply:
As discussed, it was not part of prd and design, will incorporate this later.